bindgen/TypeTranslator.cpp

```diff
@@ -68,111 +57,143 @@ TypeTranslator::TranslateFunctionPointer(const clang::QualType &qtpe,
     }
 }
 
-std::string TypeTranslator::TranslatePointer(const clang::QualType &pte,
-                                             const std::string *avoid) {
+Type *TypeTranslator::TranslatePointer(const clang::QualType &pte,
+                                       const std::string *avoid) {
 
     if (pte->isBuiltinType()) {
         const clang::BuiltinType *as = pte->getAs<clang::BuiltinType>();
 
         // Take care of void*
         if (as->getKind() == clang::BuiltinType::Void) {
-            return "native.Ptr[Byte]";
+            return new PointerType(new PrimitiveType("Byte"));
         }
 
         // Take care of char*
         if (as->getKind() == clang::BuiltinType::Char_S ||
             as->getKind() == clang::BuiltinType::SChar) {
-            return "native.CString";
+            return new PrimitiveType("native.CString");
         }
     }
 
-    return std::string("native.Ptr[") + Translate(pte, avoid) +
-           std::string("]");
+    return new PointerType(translate(pte, avoid));
 }
 
-std::string
-TypeTranslator::TranslateStructOrUnion(const clang::QualType &qtpe) {
+Type *TypeTranslator::translateStruct(const clang::QualType &qtpe) {
     if (qtpe->hasUnnamedOrLocalType()) {
         // TODO: Verify that the local part is not a problem
         uint64_t size = ctx->getTypeSize(qtpe);
-        return "native.CArray[Byte, " + uint64ToScalaNat(size) + "]";
+        return new ArrayType(new PrimitiveType("Byte"), size);
     }
 
     std::string name = qtpe.getUnqualifiedType().getAsString();
 
     // TODO: do it properly
     size_t f = name.find(std::string("struct __dirstream"));
     if (f != std::string::npos) {
-        return std::string("native.CArray[Byte, Digit[_3, Digit[_2, _0]]]");
+        return new ArrayType(new PrimitiveType("Byte"), 320);
     }
 
-    f = name.find(" ");
-    if (f != std::string::npos) {
-        return name.replace(f, std::string(" ").length(), "_");
+    auto it = aliasesMap.find(name);
+    if (it != aliasesMap.end()) {
+        /* name contains space: struct <name>.
+         * Use type alias instead struct type */
+        return (*it).second;
```

Comment:
Consider following example:
```c
struct myStruct {
    int a;
};
```
Generated Scala code for the struct:
```Scala
type struct_myStruct = native.CStruct1[native.CInt]
```
Then we find the function that uses out struct:
```c
void myFunction(struct myStruct, s);
```
`TypeTranslator` would translate `struct myStruct` into `native.CStruct1[native.CInt]` but we want to use the alias `struct_myStruct`. We know that the alias already exists because struct was defined, so we find the alias in the map.

So we will reference a `TypeDef` instance generated for the struct